### ISSUE TITLE
Bug: [no-misused-promises] `inheritedMethods` check should flag all statically analyzable declarations, not just identifiers and numbers.

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10211]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=MYewdgzgLgBNCGUCWwDKBPAtgIxAGxgF4YMd8A6AMxACcAKAcgWWDi1zwYEoBuAKD5IwUAKY1K8YCJgBJYWIlSYAbz4wYSACYjhSSkjEAuGHS5EAfDABuILf3UBGY6YvXbm%2BzABMzs4Us2dmpwUDRCAOYAMkiiNPB4vq6BHsGgmAAOAK6imqihEdGx8Yn%2BbkHqAPQVMADuIngEANZgIDVgbGR4wQDapBzkMWKItAC6JQHu-MFVMHgiEBBwmTTSejBQABZIi%2BCri%2BkgC0jYc%2Bsgs0Kw1DQ9zCh9%2BGMmfhNBAL4CfMB48AswAMI-ABeQI0GTmmB0UEWcliimkqnUv3QYFYWihegM9DMyhgHyREBRrAcLlx%2BJgyNRMG6XhGpLxwUprCY%2BTAUUGcU49PJTJgDDSWRyeTCbMKQy5OIZBKJ1IeeAGRSgo25nl53TuaHYjxVfDeQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkAO4BDaE0jgwAXxBqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const staticSymbol = Symbol.for('static symbol');

interface Interface {
  identifier: () => void;
  1: () => void;
  2: () => void;
  stringLiteral: () => void;
  computedStringLiteral: () => void;
  // well known symbol
  [Symbol.iterator]: () => void;

  // less sure if this one is possible to lint for
  [staticSymbol]: () => void;
}


class Clazz implements Interface {
  async identifier() { }
  async 1() { }
  async [2]() { }
  async 'stringLiteral'() { }
  async 'computedStringLiteral'() { }
  async [Symbol.iterator]() { };
  async [staticSymbol]() { };
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-misused-promises": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

All method declarations should flag

### Actual Result

Only `async identifier() {}` and `async 3() {}` flagged.

**Comment by kirkwaiblinger** [comment]
We should take a look at object literal assignments too, which suffer most of the same defects: 
([playground](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=MYewdgzgLgBNCGUCWwDKBPAtgIxAGxgF4YMd8A6AMxACcAKAcgWWDi1zwYEoBuAKD5IwUAKY1K8YCJgBJYWIlSYAbz4wYSACYjhSSkjEAuGHS5EAfDABuILf3UBGY6YvXbm%2BzABMzs4Us2dmpwUDRCAOYAMkiiNPB4vq6BHsGgmAAOAK6imqihEdGx8Yn%2BbkHqAPQVMADuIngEANZgIDVgbGR4wQDapBzkMWKItAC6JQHu-MFVMHgiEBBwmTTSejBQABZIi%2BCri%2BkgC0jYc%2Bsgs0Kw1DQ9zCh9%2BGMmfhNBAL4CfKCQVyAgxnJYoppMRVOp4BB0GBWFodMh9GIXMoYG8ADTBCFQ1gOJEo9HgyHQmDdLwjXFojGE1hMfJgKKDOKccn4mCYondBhpLI5PJhOmFIacMlmZEUglY4kPPADIpQUbMykS7p3NDsR4Kt5AA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkAO4BDaE0jgwAXxBqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false))
```ts
const foo: Interface = {
  async identifier() { },
  async 1() { },
  async [2]() { },
  async 'stringLiteral'() { },
  async ['computedStringLiteral']() { },
  async [Symbol.iterator]() { },
  async [staticSymbol]() { },
}
```

This one additionally gets the `'stringLiteral'` case correct though.

**Comment by bradzacher** [comment]
> `async [2]() { }`
> `async ['computedStringLiteral']() { }`

In general we ignore all computed property names in rules because they can be any expression at all.

We _could_ special case the literal cases but generally we just treat computed members as a black box that we can't truly resolve cos doing the above cases is really rare.

> `async [Symbol.iterator]() { };`

This falls into the same boat - it's a computed member so we ignore it. We could special case the `Symbol` special members but it's also a difficult thing because like `Symbol.notIterator` is valid code too - so would we special case it or ignore it?

Then you would also wanna handle and ignore like
```ts
const Symbol = {iterator: 1};
class Foo {
  [Symbol.iterator](){}
}
```
(people do like to reuse the name `Symbol` more than you'd like - eg the TS codebase itself!)

> `async [staticSymbol]() { };`

Same case again except now the general symbol case. You'd need to check - idk if it creates a unique type which allows us to uniquely identify the symbol. Regardless it's not possible to check this case without type info which does increase the cost of the check. Often not _super_ worth checking this either as it's really really rare to use a unique symbol in an inheritance case like this.

**Comment by bradzacher** [comment]
We should probably build some primitives around this more generally across all rules to allow us to get a property key handling the simple computed cases and similar special named cases like iterator.

**Comment by kirkwaiblinger** [comment]
yeah, we do have `getStaticMemberAccessValue`, which I think should handle most if not all of these already. Not super certain if it will get us the well-known symbols, but I know it handles things like `Symbol.for('foo')` 🤷
